tests/test_parse.py

```diff
@@ -151,3 +151,25 @@ class Model(BaseModel):
 
     m = Model.parse_obj({'a': {'key': 'B', 'foo': 3}})
     assert isinstance(m.a, SubModelB)
+
+
+def test_parse_nested_custom_root_mapping():
+    class NestedRootMapping(BaseModel):
+        __root__: Dict[int, 'NestedRootMapping']
+
+    NestedRootMapping.update_forward_refs()
+
+    obj = NestedRootMapping.parse_obj({'1': {'2': {}}})
+    assert obj.dict() == {'__root__': {1: {'__root__': {2: {'__root__': {}}}}}}
+    assert obj.json() == '{"1": {"2": {}}}'
+
+
+def test_parse_nested_custom_root_tuple():
+    class NestedRootTuple(BaseModel):
+        __root__: Tuple[int, Optional["NestedRootTuple"]]
+
+    NestedRootTuple.update_forward_refs()
```

Comment:
I discovered a strange bug here -- if you don't call this `update_forward_refs`, it doesn't raise an error, but then below you actually get:

```python
    obj = NestedRootTuple.parse_obj(('1', ('2', ('3', None))))
    assert obj.dict() == {'__root__': (1, {'__root__': ('2', {'__root__': ('3', None)})})}  # note 2 and 3 became strings
```

Might be worth looking into separately.

Reply:
Yep, I confirmed this is not related to custom root; creating a separate issue.